pkg/process/process.go

```diff
@@ -0,0 +1,143 @@
+// Copyright 2022-2023 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package process
+
+import (
+	"fmt"
+
+	"github.com/prometheus/procfs"
+
+	"github.com/parca-dev/parca-agent/pkg/cgroup"
+)
+
+// key is a unique identifier for a process.
+// TODO(kakkoyun): This is not unique enough. We will need to add more fields.
+// For example, we can use the process start time to make sure we are not.
+// However, reading the stats file is expensive. For now, we will rely on the PID.
+// I will keep the struct to make it easier to add more fields in the future.
+type key struct {
+	pid int
+}
+
+type psTree map[key]*process
+
+type process struct {
+	procfs.Proc
+
+	tree *psTree
+
+	parent    int
+	starttime uint64
+}
+
+func (p *process) ancestors() []*procfs.Proc {
+	ancestors := []*procfs.Proc{}
+	proc := p
+	for {
+		if proc.parent == 0 {
+			break
+		}
+		k := key{pid: proc.parent}
+		parent, ok := (*p.tree)[k]
+		if !ok {
+			break
+		}
+		ancestors = append(ancestors, &parent.Proc)
+		proc = parent
+	}
+	return ancestors
+}
+
+// TODO(kakkoyun): This is an ever growing map. Introduce a mechanism to prune it.
```

Comment:
I think it would be better to not have this be a package global, but rather a struct that's instantiated once and passed around. Then lifecycle management with a long-running go-routine is also easier.

Reply:
I'll do this. This was hard to commit for me anyway 😅